### PR TITLE
Update documentation and deprecation warnings for split classes

### PR DESCRIPTION
**Pull request by big-guy**
### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/.github/CONTRIBUTING.md)
- [ ] [Sign Gradle CLA](http://gradle.org/contributor-license-agreement/)
- [ ] [Link to Design Spec](https://github.com/gradle/gradle/tree/master/design-docs) for changes that affect more than 1 public API (that is, not in an `internal` package) or updates to > 20 files
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass locally: `./gradlew quickCheck <impacted-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation including proper use of `@since` and `@Incubating` annotations for all public APIs
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by big-guy** [comment]
@bmuschko could you take a look at this?

File: subprojects/docs/src/docs/release/notes.md [hunk 1]
```diff
@@ -355,21 +355,39 @@ To return to the old behavior, explicitly set the classes directory:
 
 Please be aware that this will interfere with the effectiveness of the build cache when using multiple JVM languages in the same source set. Gradle will disable caching for tasks when it detects that multiple tasks create outputs in the same location.
 
+### Detecting test classes for custom `Test` tasks
+
+Before Gradle 4.0, all test classes were compiled into a single output directory. As [described above](#multiple-class-directories-for-a-single-source-set), Gradle now uses separate classes directories for each language in a source set. 
+
+Builds that define custom `Test` tasks may not find the same test classes due to these changes if tests are written in languages other than Java. Deprecation warnings warn about this behavior.
```
Comment (bmuschko) [inline subprojects/docs/src/docs/release/notes.md]: I think we'll have to explicitly point out that tests are not run if the deprecation message is not fixed.

File: subprojects/plugins/src/integTest/groovy/org/gradle/api/plugins/JavaPluginIntegrationTest.groovy [hunk 2]
```diff
@@ -61,6 +61,19 @@ class JavaPluginIntegrationTest extends AbstractIntegrationSpec {
         then:
         file("build/classes/java/main").assertDoesNotExist()
         file("build/classes/main/Main.class").assertExists()
-        result.assertOutputContains("Using a single directory for all classes from a source set.")
+    }
+
+    def "emits deprecation message if something uses classesDir"() {
```
Comment (bmuschko) [inline subprojects/plugins/src/integTest/groovy/org/gradle/api/plugins/JavaPluginIntegrationTest.groovy]: Not sure if we need to test that but the `test` task would also have the status `NO-SOURCE`.

File: subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java [hunk 3]
```diff
@@ -906,6 +906,7 @@ public Test setTestNameIncludePatterns(List<String> testNamePattern) {
      * Returns the root folder for the compiled test sources.
```
Comment (bmuschko) [inline subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java]: Noticed on the side lines: `setTestClassesDirs` does not add `@Incubating`.

File: subprojects/plugins/src/main/java/org/gradle/api/internal/tasks/DefaultSourceSetOutput.java [hunk 4]
```diff
@@ -76,7 +76,7 @@ public File getClassesDir() {
         if (isLegacyLayout()) {
             return fileResolver.resolve(classesDir);
         }
-        SingleMessageLogger.nagUserOfDeprecatedBehaviour("Using a single directory for all classes from a source set");
+        SingleMessageLogger.nagUserOfDeprecatedBehaviour("Gradle now uses separate output directories for each JVM language, but this build assumes a single directory for all classes from a source set");
```
Comment (bmuschko) [inline subprojects/plugins/src/main/java/org/gradle/api/internal/tasks/DefaultSourceSetOutput.java]: Maybe needs to indicate that `Test` task action will be skipped.

**Pull request merged**